File: com.ca.lsp.cobol/lsp-core-cobol-parser/src/main/java/com/ca/lsp/core/cobol/parser/listener/Listener.java
/*
 * Copyright (c) 2019 Broadcom.
 * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Broadcom, Inc. - initial API and implementation
 */
package com.ca.lsp.core.cobol.parser.listener;

import com.broadcom.lsp.domain.cobol.event.model.Position;
import com.ca.lsp.core.cobol.model.SyntaxError;
import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.List;
import java.util.Optional;

@AllArgsConstructor
public abstract class Listener {
  private static final int PREPROCESSING_ERROR_INDEX = -1;
  @Getter private List<SyntaxError> errorsPipe;

  public void syntaxError(
      String documentName,
      int line,
      int charPositionInLine,
      String msg,
      int errorLength,
      int severity) {
    registerError(
        msg,
        new Position(
            documentName,
            PREPROCESSING_ERROR_INDEX,
            charPositionInLine,
            (charPositionInLine + errorLength),
            line,
            charPositionInLine),
        severity);
  }

  public void syntaxError(
      String documentName,
      int line,
      int charPositionInLine,
      int charEndingIndex,
      String msg,
      int severity) {
    registerError(
        msg,
        new Position(
            documentName,
            PREPROCESSING_ERROR_INDEX,
            charPositionInLine,
            charEndingIndex,
            line,
            charPositionInLine),
        severity);
  }

  public void unregisterError(int line) {
    Optional<SyntaxError> error =
        getErrorsPipe().stream().filter(err -> err.getPosition().getLine() == line).findFirst();
    error.ifPresent(err -> getErrorsPipe().remove(err));
  }

  void registerError(String msg, Position position, int severity) {
    if (getErrorsPipe() != null) {
      getErrorsPipe()
          .add(
              SyntaxError.syntaxerror()
                  .position(position)
                  .suggestion(msg)
                  .type(0)
                  .severity(severity)
                  .build());
    }
  }
}

File: clients/analysis/src/vm/listener.ts
/*
 * Copyright (c) 2025 Broadcom.
 * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Broadcom, Inc. - initial API and implementation
 */
import { Program, Location, Paragraph, Section } from "../model/cfast";
import { CobolInstruction } from "./instructions";

/**
 * Virtual Processor Listener is used to obtain information during COBOL statements processing
 */
export interface VirtualProcessorListener {
  /**
   * Fires wnen Virtual Processor tryes to evaluate not supported COBOL statement
   * @param location the COBOL statement location
   */
  notSupported(location: Location): void;

  /**
   * Fires when control moves from previous node to the next node
   * @param node1 - previous node
   * @param node2 - next node
   */
  moveControl(
    node1: Paragraph | Section | Program,
    node2: Paragraph | Section | Program,
  ): void;

  /**
   * Fires when Virtual Processor reach the maximum VM count
   */
  maximumVMCountReached(): void;

  /**
   * Fires when latest statement of the program was executed and control falls down
   * @param path a VM path as a array of executed cobol instructions
   */
  reportFallThru(path: CobolInstruction[]): void;
}
